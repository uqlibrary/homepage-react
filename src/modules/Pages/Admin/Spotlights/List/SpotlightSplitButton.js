import React from 'react';
import PropTypes from 'prop-types';

import ArrowDropDownIcon from '@mui/icons-material/ArrowDropDown';
import Button from '@mui/material/Button';
import ButtonGroup from '@mui/material/ButtonGroup';
import ClickAwayListener from '@mui/material/ClickAwayListener';
import Grid from '@mui/material/Grid';
import Grow from '@mui/material/Grow';
import Paper from '@mui/material/Paper';
import Popper from '@mui/material/Popper';
import MenuItem from '@mui/material/MenuItem';
import MenuList from '@mui/material/MenuList';

import { useConfirmationState } from 'hooks';

import { ConfirmationBox } from 'modules/SharedComponents/Toolbox/ConfirmDialogBox';
import { default as locale } from 'modules/Pages/Admin/Spotlights/spotlightsadmin.locale';
import { isPastSpotlight, navigateToEditForm } from 'modules/Pages/Admin/Spotlights/spotlighthelpers';

// based on https://material-ui.com/components/button-group/ "Split button"
export const SpotlightSplitButton = ({
    deleteSpotlightById,
    mainButtonLabel = 'Edit',
    navigateToCloneForm,
    navigateToView,
    confirmDeleteLocale = false,
    showViewByHistoryOption,
    spotlight,
    history,
    allowedArrowActions = [],
}) => {
    const [open, setOpen] = React.useState(false);
    const anchorRef = React.useRef(null);
    const [isDeleteConfirmOpen, showDeleteConfirmation, hideDeleteConfirmation] = useConfirmationState();

    const handleToggle = () => {
        setOpen(prevOpen => !prevOpen);
    };

    const handleClose = event => {
        /* istanbul ignore next */
        if (anchorRef.current && anchorRef.current.contains(event.target)) {
            return;
        }

        setOpen(false);
    };

    function deleteLocale() {
        // when delete isnt supplied, we have to not let the function fire
        if (allowedArrowActions.includes('delete') && typeof confirmDeleteLocale === 'function') {
            return confirmDeleteLocale(1);
        }
        return {
            ...locale.listPage.confirmDelete,
            confirmationTitle: 'should not appear',
        };
    }

    return (
        <React.Fragment>
            <ConfirmationBox
                actionButtonColor="secondary"
                actionButtonVariant="contained"
                confirmationBoxId="spotlight-delete-confirm"
                onAction={() => deleteSpotlightById(spotlight.id)}
                onClose={hideDeleteConfirmation}
                onCancelAction={hideDeleteConfirmation}
                isOpen={isDeleteConfirmOpen}
                locale={deleteLocale()}
            />
            <Grid container direction="column" alignItems="center">
                <Grid item xs={12} sx={{ position: 'relative', minHeight: 50 }}>
                    <ButtonGroup
                        variant="contained"
                        color="primary"
                        ref={anchorRef}
                        aria-label="split button"
                        style={{ float: 'right' }}
                    >
                        {mainButtonLabel === 'Edit' && (
                            <Button
                                children={mainButtonLabel}
                                color="primary"
                                data-testid={`spotlight-list-item-edit-${spotlight.id}`}
                                id={`spotlight-list-item-edit-${spotlight.id}`}
                                onClick={() => navigateToEditForm(spotlight.id, history)}
                                className={'editButton'}
                                variant="contained"
                            />
                        )}
                        {mainButtonLabel === 'View' && (
                            <Button
                                children={mainButtonLabel}
                                color="primary"
                                data-testid={`spotlight-list-item-view-${spotlight.id}`}
                                id={`spotlight-list-item-view-${spotlight.id}`}
                                onClick={() => navigateToView(spotlight.id, history)}
                                className={'editButton'}
                                variant="contained"
                            />
                        )}
                        {mainButtonLabel === 'Clone' && (
                            <Button
                                children={mainButtonLabel}
                                color="primary"
                                data-testid={`spotlight-list-item-clone-${spotlight.id}`}
                                id={`spotlight-list-item-clone-${spotlight.id}`}
                                onClick={() => navigateToCloneForm(spotlight.id, history)}
                                className={'editButton'}
                                variant="contained"
                            />
                        )}
                        {allowedArrowActions.length > 0 && (
                            <Button
                                color="primary"
                                className={'editButton'}
                                size="small"
                                aria-controls={open ? 'split-button-menu' : undefined}
                                aria-expanded={open ? 'true' : undefined}
                                aria-label="More actions"
                                aria-haspopup="menu"
                                data-testid={`spotlight-list-arrowicon-${spotlight.id}`}
                                onClick={handleToggle}
                                title="More actions"
                            >
                                <ArrowDropDownIcon />
                            </Button>
                        )}
                    </ButtonGroup>
                    <Popper open={open} anchorEl={anchorRef.current} role={undefined} transition disablePortal>
                        {({ TransitionProps, placement }) => (
                            <Grow
                                {...TransitionProps}
                                style={{
                                    transformOrigin:
                                        placement === 'bottom'
                                            ? 'center top'
                                            : /* istanbul ignore next */ 'center bottom',
                                    zIndex: 1,
                                    minWidth: 60,
                                }}
                            >
                                <Paper sx={{ marginTop: '-75px' }}>
                                    <ClickAwayListener onClickAway={handleClose}>
                                        <MenuList id="split-button-menu">
                                            {allowedArrowActions.includes('clone') && (
                                                <MenuItem
                                                    data-testid={`${spotlight.id}-clone-button`}
                                                    key={`${spotlight.id}-clone-button`}
                                                    onClick={() => navigateToCloneForm(spotlight.id, history)}
                                                >
                                                    {locale.form.splitButton.labels.clone}
                                                </MenuItem>
                                            )}
                                            {allowedArrowActions.includes('delete') && (
                                                <MenuItem
                                                    data-testid={`${spotlight.id}-delete-button`}
                                                    key={`${spotlight.id}-delete-button`}
                                                    onClick={showDeleteConfirmation}
                                                >
                                                    {locale.form.splitButton.labels.delete}
                                                </MenuItem>
                                            )}
                                            {allowedArrowActions.includes('viewbyhistory') && (
                                                <MenuItem
                                                    data-testid={`${spotlight.id}-viewbyhistory-button`}
                                                    key={`${spotlight.id}-viewbyhistory-button`}
                                                    onClick={() =>
                                                        !!showViewByHistoryOption && showViewByHistoryOption(spotlight)
                                                    }
                                                >
                                                    {locale.form.splitButton.labels.history}
                                                </MenuItem>
                                            )}
                                            {mainButtonLabel !== 'Edit' &&
                                                allowedArrowActions.includes('edit') &&
                                                !isPastSpotlight(spotlight) && (
                                                    /* istanbul ignore next */ <MenuItem
                                                        data-testid={`${spotlight.id}-viewbyhistory-button`}
                                                        key={`${spotlight.id}-viewbyhistory-button`}
                                                        onClick={
                                                            /* istanbul ignore next */ () =>
                                                                /* istanbul ignore next */
                                                                navigateToEditForm(spotlight.id, history)
                                                        }
                                                    >
                                                        {locale.form.splitButton.labels.edit}
                                                    </MenuItem>
                                                )}
                                            {mainButtonLabel !== 'View' &&
                                                allowedArrowActions.includes('view') &&
                                                isPastSpotlight(spotlight) && (
                                                    /* istanbul ignore next */ <MenuItem
                                                        data-testid={`${spotlight.id}-viewbyhistory-button`}
                                                        key={`${spotlight.id}-viewbyhistory-button`}
                                                        onClick={
                                                            /* istanbul ignore next */ () =>
                                                                /* istanbul ignore next */ navigateToView(
                                                                    spotlight.id,
                                                                    history,
                                                                )
                                                        }
                                                    >
                                                        {locale.form.splitButton.labels.view}
                                                    </MenuItem>
                                                )}
                                        </MenuList>
                                    </ClickAwayListener>
                                </Paper>
                            </Grow>
                        )}
                    </Popper>
                </Grid>
            </Grid>
        </React.Fragment>
    );
};

SpotlightSplitButton.propTypes = {
    spotlight: PropTypes.any,
    mainButtonLabel: PropTypes.string,
    deleteSpotlightById: PropTypes.func,
    navigateToCloneForm: PropTypes.func,
    navigateToView: PropTypes.func,
    navigateToMainFunction: PropTypes.func,
    confirmDeleteLocale: PropTypes.any,
    showViewByHistoryOption: PropTypes.func,
    history: PropTypes.any,
    allowedArrowActions: PropTypes.array,
};

export default SpotlightSplitButton;
