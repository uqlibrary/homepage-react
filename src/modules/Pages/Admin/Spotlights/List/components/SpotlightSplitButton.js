import React from 'react';
import PropTypes from 'prop-types';

import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import Button from '@material-ui/core/Button';
import ButtonGroup from '@material-ui/core/ButtonGroup';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Grid from '@material-ui/core/Grid';
import Grow from '@material-ui/core/Grow';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import { makeStyles } from '@material-ui/core/styles';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';

import { useConfirmationState } from 'hooks';

import { ConfirmationBox } from 'modules/SharedComponents/Toolbox/ConfirmDialogBox';

// based on https://material-ui.com/components/button-group/ "Split button"

const useStyles = makeStyles(() => ({
    parent: {
        position: 'relative',
        minHeight: 50,
    },
    menuWrapper: {
        marginTop: -50,
    },
}));
export const SpotlightSplitButton = ({
    spotlightId,
    deleteSpotlightById,
    mainButtonLabel,
    navigateToCloneForm,
    navigateToEditForm,
    navigateToView,
    confirmDeleteLocale,
}) => {
    const classes = useStyles();
    const [open, setOpen] = React.useState(false);
    const anchorRef = React.useRef(null);
    const [isDeleteConfirmOpen, showDeleteConfirmation, hideDeleteConfirmation] = useConfirmationState();

    const handleToggle = () => {
        setOpen(prevOpen => !prevOpen);
    };

    const handleClose = event => {
        /* istanbul ignore next */
        if (anchorRef.current && anchorRef.current.contains(event.target)) {
            return;
        }

        setOpen(false);
    };

    return (
        <React.Fragment>
            <ConfirmationBox
                actionButtonColor="secondary"
                actionButtonVariant="contained"
                confirmationBoxId="spotlight-delete-confirm"
                onAction={() => deleteSpotlightById(spotlightId)}
                onClose={hideDeleteConfirmation}
                onCancelAction={hideDeleteConfirmation}
                isOpen={isDeleteConfirmOpen}
                locale={confirmDeleteLocale(1)}
            />
            <Grid container direction="column" alignItems="center">
                <Grid item xs={12} className={classes.parent}>
                    <ButtonGroup variant="contained" color="primary" ref={anchorRef} aria-label="split button">
                        <Button
                            children={mainButtonLabel}
                            color="primary"
                            data-testid={`spotlight-list-item-${mainButtonLabel.toLowerCase()}-${spotlightId}`}
                            id={`spotlight-list-item-${mainButtonLabel.toLowerCase()}-${spotlightId}`}
                            onClick={() =>
                                mainButtonLabel === 'Edit'
                                    ? navigateToEditForm(spotlightId)
                                    : navigateToView(spotlightId)
                            }
                            className={classes.editButton}
                            variant="contained"
                        />
                        <Button
                            color="primary"
                            className={classes.editButton}
                            size="small"
                            aria-controls={open ? 'split-button-menu' : undefined}
                            aria-expanded={open ? 'true' : undefined}
                            aria-label="More actions"
                            aria-haspopup="menu"
                            data-testid={`spotlight-list-arrowicon-${spotlightId}`}
                            onClick={handleToggle}
                            title="More actions"
                        >
                            <ArrowDropDownIcon />
                        </Button>
                    </ButtonGroup>
                    <Popper open={open} anchorEl={anchorRef.current} role={undefined} transition disablePortal>
                        {({ TransitionProps, placement }) => (
                            <Grow
                                {...TransitionProps}
                                style={{
                                    transformOrigin:
                                        placement === 'bottom'
                                            ? 'center top'
                                            : /* istanbul ignore next */ 'center bottom',
                                    zIndex: 1,
                                }}
                            >
                                <Paper className={classes.menuWrapper}>
                                    <ClickAwayListener onClickAway={handleClose}>
                                        <MenuList id="split-button-menu">
                                            <MenuItem
                                                data-testid={`${spotlightId}-clone-button`}
                                                key={`${spotlightId}-clone-button`}
                                                onClick={() => navigateToCloneForm(spotlightId)}
                                            >
                                                Clone
                                            </MenuItem>
                                            <MenuItem
                                                data-testid={`${spotlightId}-delete-button`}
                                                key={`${spotlightId}-delete-button`}
                                                onClick={showDeleteConfirmation}
                                            >
                                                Delete
                                            </MenuItem>
                                        </MenuList>
                                    </ClickAwayListener>
                                </Paper>
                            </Grow>
                        )}
                    </Popper>
                </Grid>
            </Grid>
        </React.Fragment>
    );
};

SpotlightSplitButton.propTypes = {
    spotlightId: PropTypes.string,
    mainButtonLabel: PropTypes.string,
    deleteSpotlightById: PropTypes.func,
    navigateToCloneForm: PropTypes.func,
    navigateToEditForm: PropTypes.func,
    navigateToView: PropTypes.func,
    navigateToMainFunction: PropTypes.func,
    confirmDeleteLocale: PropTypes.func,
};

SpotlightSplitButton.defaultProps = {
    mainButtonLabel: 'Edit',
};

export default SpotlightSplitButton;
