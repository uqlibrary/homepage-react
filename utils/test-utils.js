/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import preview from 'jest-preview';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
    return render(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const RenderWithRouter = ({
    children,
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
}) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

// eslint-disable-next-line react/prop-types
export const WithRedux = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>{children}</Provider>
);

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    WithRedux,
    WithReduxStore,
    RenderWithRouter,
    preview,
};
