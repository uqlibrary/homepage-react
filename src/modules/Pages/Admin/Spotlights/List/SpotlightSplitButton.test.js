import React from 'react';
import { rtlRender } from 'test-utils';
import SpotlightSplitButton from './SpotlightSplitButton';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => jest.fn(),
}));

function setup(testProps = {}) {
    const props = {
        deleteSpotlightById: jest.fn(),
        mainButtonLabel: 'Edit',
        navigateToCloneForm: jest.fn(),
        navigateToView: jest.fn(),
        confirmDeleteLocale: jest.fn(),
        showViewByHistoryOption: jest.fn(),
        spotlight: [],
    };
    return rtlRender(<SpotlightSplitButton {...props} {...testProps} />);
}

describe('SpotlightSplitButton', () => {
    it('default values render correctly', () => {
        const { container } = setup();
        expect(container).toMatchSnapshot();
    });
});
