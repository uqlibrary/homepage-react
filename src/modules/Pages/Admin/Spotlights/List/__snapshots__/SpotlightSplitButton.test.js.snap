// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotlightSplitButton default values render correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-pq3v65-MuiGrid-root"
    >
      <div
        aria-label="split button"
        class="MuiButtonGroup-root MuiButtonGroup-contained css-1mtivgc-MuiButtonGroup-root"
        role="group"
        style="float: right;"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary editButton css-t41e4y-MuiButtonBase-root-MuiButton-root"
          data-testid="spotlight-list-item-edit-undefined"
          id="spotlight-list-item-edit-undefined"
          tabindex="0"
          type="button"
        >
          Edit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
